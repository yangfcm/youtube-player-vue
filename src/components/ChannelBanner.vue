<script setup lang="ts">
const props = defineProps<{
  imageUrl?: string
}>()

</script>

<template>
  <template v-if="imageUrl">
    <div class="app-channel-banner" :style="`background-image: url(${props.imageUrl})`">
    </div>
  </template>
</template>

<style>
  .app-channel-banner {
    /* background-image: "url(v-bind('bannerImageUrl'))"; */
    width: 100%;
    height: 16vw;
    background-size: cover;
    background-repeat: no-repeat;
    background-position: 0 50%;
  }
</style>