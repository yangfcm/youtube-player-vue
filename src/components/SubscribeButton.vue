<script setup lang="ts">
import { ref } from 'vue'
import { useSubscribe } from '@/composables/useSubscribe';
import { useAuth } from '@/composables/useAuth';

const props = defineProps<{
  channelId: string
}>()

const hovered = ref(false)
const { isSignedIn } = useAuth()
const { status, error, subscribed, subscribeChannel, unsubscribeChannel } = useSubscribe(props.channelId)


</script>

<template>
  <v-btn
    v-if="isSignedIn && subscribed !== undefined"
    :color="subscribed ? 'secondary' : ''"
    width="150"
    @mouseenter="() => hovered = true"
    @mouseleave="() => hovered = false"
  >
    {{ subscribed ? 
      (hovered ? 'Unsubscribe' : 'Subscribed') : 
      'Subscribe' }}
  </v-btn>
</template>