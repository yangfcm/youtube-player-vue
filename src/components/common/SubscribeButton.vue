<template>
  <div style="padding-top: 8px;">
    <div v-if="signedIn===true">
      <button
        class="ui inverted button"
        :class="subscriptionId ? 'red' : 'orange'"
        style="width: 120px;"
        @mouseover="handleMouseOver"
        @mouseleave="handleMouseLeave"
        @click="handleToggleSubscribe"
        :disabled="buttonText==='Subscribing' || buttonText==='Unsubscribing'"
      >{{ buttonText }}</button>
    </div>
  </div>
</template>

<script>
import { mapState, mapGetters, mapActions } from "vuex";

export default {
  data: function () {
    return {
      subscriptionId: "",
      buttonText: "",
      BUTTON_TEXT: {
        UNSUBSCRIBE: "Unsubscribe",
        SUBSCRIBE: "Subscribe",
        SUBSCRIBING: "Subscribing",
        UNSUBSCRIBING: "Unsubscribing",
        SUBSCRIBED: "Subscribed",
        UNSUBSCRIBED: "Unsubscribed",
      },
    };
  },
  props: ["channelId"],
  computed: {
    ...mapState(["channel"]),
    ...mapGetters(["signedIn"]),
  },
  watch: {
    signedIn(value) {
      this.buttonText = "";
      if (value) {
        this.fetchChannelSubscriptionId();
      } else {
        this.subscriptionId = "";
      }
    },
  },
  methods: {
    ...mapActions([
      "subscribeChannel",
      "unsubscribeChannel",
      "fetchChannelSubscription",
    ]),

    handleMouseOver() {
      if (this.buttonText === "Subscribed") {
        this.buttonText = this.BUTTON_TEXT.UNSUBSCRIBE;
      }
      if (this.buttonText === "Unsubscribed") {
        this.buttonText = this.BUTTON_TEXT.SUBSCRIBE;
      }
    },

    handleMouseLeave() {
      if (this.buttonText === "Subscribe") {
        this.buttonText = this.BUTTON_TEXT.UNSUBSCRIBED;
      }
      if (this.buttonText === "Unsubscribe") {
        this.buttonText = this.BUTTON_TEXT.SUBSCRIBED;
      }
    },

    async handleToggleSubscribe() {
      if (this.subscriptionId) {
        // Unsubscribe channel
        this.buttonText = this.BUTTON_TEXT.UNSUBSCRIBING;
        await this.unsubscribeChannel(this.subscriptionId);
      } else {
        // Subscribe channel
        this.buttonText = this.BUTTON_TEXT.SUBSCRIBING;
        await this.subscribeChannel(this.channelId);
      }
      this.subscriptionId = this.channel.subscriptionId;
      // When subscribe/unsubscribe operation finishes, update button text
      if (!this.subscriptionId) {
        this.buttonText = this.BUTTON_TEXT.UNSUBSCRIBED;
      } else {
        this.buttonText = this.BUTTON_TEXT.SUBSCRIBED;
      }
    },

    fetchChannelSubscriptionId() {
      this.fetchChannelSubscription(this.channelId).then(() => {
        this.subscriptionId = this.channel.subscriptionId;
        if (this.subscriptionId) {
          this.buttonText = this.BUTTON_TEXT.SUBSCRIBED;
        } else {
          this.buttonText = this.BUTTON_TEXT.UNSUBSCRIBED;
        }
      });
    },
  },

  created() {
    this.fetchChannelSubscriptionId();
  },
};
</script>

<style scoped>
</style>