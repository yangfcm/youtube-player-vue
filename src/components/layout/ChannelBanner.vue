<template>
  <div class="app-banner-container" v-if="channelIntro">
    <div class="app-banner-title-container">
      <div class="app-banner-image">
        <img
          class="ui tiny image circular"
          :src="channelIntro.snippet.thumbnails.medium.url"
          :alt="channelIntro.snippet.title"
        />
      </div>
      <div class="app-banner-title">
        <h2 class="ui header app-banner-title-text">{{ channelIntro.snippet.title }}</h2>
      </div>
    </div>
    <!-- <div>
        <app-subscribe-button-auth
          [channelId]="channelIntro.id"
        ></app-subscribe-button-auth>
    </div>-->
  </div>
</template>

<script>
export default {
  props: ["channelIntro"],
};
</script>

<style scoped>
.app-banner-container {
  margin-top: 1.5rem;
  margin-bottom: 1.5rem;
  display: flex;
  justify-content: space-between;
  align-items: center;
}
.app-banner-title-container {
  display: flex;
}
.app-banner-image {
  margin-right: 1rem;
  min-height: 80px;
  display: flex;
  align-items: center;
}
.app-banner-title {
  display: flex !important;
  align-items: center;
}
.app-banner-title-text {
  display: flex;
  padding-top: 5px;
  padding-bottom: 5px;
}
@media (max-width: 576px) {
  .app-banner-title-text {
    flex-direction: column;
  }
}
@media (max-width: 576px) {
  .app-banner-container {
    flex-direction: column;
  }
}
</style>