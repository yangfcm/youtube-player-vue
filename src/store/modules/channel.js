import axios from "../../settings";
import {
  FETCH_MY_CHANNELS,
  FETCH_CHANNEL_SUBSCRIPTION,
  FETCH_CHANNEL_INTRO,
  SUBSCRIBE_CHANNEL,
  UNSUBSCRIBE_CHANNEL,
  CATCH_CHANNEL_ERROR,
} from "../types";
import parseError from "../helpers/parseError";

const state = {
  myChannels: null, // channels subscribed by authed user
  channelIntro: null,
  subscriptionId: null, // Give a channel id, record its subscription id(if subscribed), or empty string(not subscribed)
  channelError: null,
};

const getters = {
  channelErrorMessage: (state) =>
    parseError(state.channelError, "Failed to fetch channel"),
};

const mutations = {
  FETCH_MY_CHANNELS: (state, payload) => {
    state.myChannels = payload;
    state.channelError = null;
  },
  FETCH_CHANNEL_SUBSCRIPTION: (state, payload) => {
    state.subscriptionId = payload;
  },
  FETCH_CHANNEL_INTRO: (state, payload) => {
    state.channelIntro = payload;
    state.channelError = null;
  },
  SUBSCRIBE_CHANNEL: (state, payload) => {
    state.subscriptionId = payload; // If a channel is subscribed successfully, save its subscription id
  },
  UNSUBSCRIBE_CHANNEL: (state) => {
    state.subscriptionId = ""; // If a subscribed channel is unsubscribed, set its subscriptionId as empty string
  },
  CATCH_CHANNEL_ERROR: (state, payload) => {
    state.channelError = payload;
  },
};

const actions = {
  /** Get the channels subscribed by authed user */
  fetchMyChannels: async (context, pageToken) => {
    try {
      const accessToken = localStorage.getItem("access_token");
      const response = await axios.get("/subscriptions", {
        headers: {
          Authorization: accessToken,
        },
        params: {
          ...axios.defaults.params,
          part: "snippet",
          maxResults: 50,
          order: "alphabetical",
          mine: true,
          pageToken,
        },
      });
      context.commit(FETCH_MY_CHANNELS, response.data);
    } catch (err) {
      context.commit(CATCH_CHANNEL_ERROR, err);
    }
  },

  /** Get the channel info by id */
  fetchChannelIntro: async (context, channelId) => {
    try {
      const response = await axios.get("/channels", {
        params: {
          ...axios.defaults.params,
          part: "snippet,statistics",
          id: channelId,
        },
      });
      context.commit(FETCH_CHANNEL_INTRO, response.data);
    } catch (err) {
      context.commit(CATCH_CHANNEL_ERROR, err);
    }
  },

  /** Check if the channel with channel id is subscribed by authed user */
  fetchChannelSubscription: async (context, channelId) => {
    try {
      const accessToken = localStorage.getItem("access_token");
      const response = await axios.get("/subscriptions", {
        headers: {
          Authorization: accessToken,
        },
        params: {
          ...axios.defaults.params,
          part: "snippet",
          forChannelId: channelId,
          mine: true,
        },
      });
      const subscribedChannel = response.data.items[0];
      context.commit(
        FETCH_CHANNEL_SUBSCRIPTION,
        subscribedChannel ? subscribedChannel.id : ""
      );
    } catch (err) {
      context.commit(CATCH_CHANNEL_ERROR, err);
    }
  },

  /** Subscribe a channel */
  subscribeChannel: async (context, channelId) => {
    try {
      const accessToken = localStorage.getItem("access_token");
      const response = await axios.post(
        "/subscriptions",
        {
          snippet: {
            resourceId: {
              kind: "youtube#channel",
              channelId,
            },
          },
        },
        {
          headers: {
            Authorization: accessToken,
          },
          params: {
            part: "snippet",
          },
        }
      );
      context.commit(SUBSCRIBE_CHANNEL, response.data.id);
    } catch (err) {
      context.commit(CATCH_CHANNEL_ERROR, err);
    }
  },

  /** Unsubscribe a channel */
  unsubscribeChannel: async (context, subscriptionId) => {
    try {
      const accessToken = localStorage.getItem("access_token");
      // const { subscriptionId } = context.state;
      await axios.delete("/subscriptions", {
        headers: {
          Authorization: accessToken,
        },
        params: {
          id: subscriptionId,
        },
      });
      context.commit(UNSUBSCRIBE_CHANNEL);
    } catch (err) {
      context.commit(CATCH_CHANNEL_ERROR, err);
    }
  },
};

export default {
  state,
  getters,
  mutations,
  actions,
};
