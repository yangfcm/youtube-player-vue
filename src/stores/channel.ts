import { ref } from 'vue'
import { defineStore } from 'pinia'
import { AsyncStatus } from '@/settings/types'
import { fetchChannelProfileAPI, fetchChannelVideosAPI, fetchChannelPlayListsAPI } from './api'
import type { ChannelDetails, PlayListsRespone, VideosResponse } from './types'

type ChannelStore = {
  profile: {
    status: AsyncStatus
    error: string
    data: Record<string, ChannelDetails>
  }
  videos: {
    status: AsyncStatus
    error: string
    data: Record<string, VideosResponse>
  }
  playlists: {
    status: AsyncStatus
    error: string
    data: Record<string, PlayListsRespone>
  }
}

export const useChannelStore = defineStore('channel', () => {
  const channel = ref<ChannelStore>({
    profile: {
      status: AsyncStatus.IDLE,
      error: '',
      data: {},
    },
    videos: {
      status: AsyncStatus.IDLE,
      error: '',
      data: {},
    },
    playlists: {
      status: AsyncStatus.IDLE,
      error: '',
      data: {},
    },
  })

  const fetchChannelProfile = async (channelId: string) => {
    try {
      channel.value.profile.status = AsyncStatus.LOADING
      channel.value.profile.error = ''
      const response = await fetchChannelProfileAPI(channelId)
      channel.value.profile.status = AsyncStatus.SUCCESS
      channel.value.profile.data[channelId] = response.data.items[0]
    } catch (err: any) {
      channel.value.profile.status = AsyncStatus.FAIL
      channel.value.profile.error = err.message
    }
  }
  const fetchChannelVideos = async (channelId: string, pageToken?: string) => {
    try {
      channel.value.videos.status = AsyncStatus.LOADING
      channel.value.videos.error = ''
      const options: Record<string, string> = {}
      if (pageToken) options.pageToken = pageToken
      const response = await fetchChannelVideosAPI(channelId, options)
      const currentItems = channel.value.videos.data[channelId]?.items || []
      channel.value.videos.data[channelId] = {
        ...response.data,
        items: [...currentItems, ...response.data.items],
      }
      channel.value.videos.status = AsyncStatus.SUCCESS
    } catch (err: any) {
      channel.value.videos.status = AsyncStatus.FAIL
      channel.value.videos.error = err.message
    }
  }
  const fetchChannelPlayLists = async (channelId: string, pageToken?: string) => {
    try {
      channel.value.playlists.status = AsyncStatus.LOADING
      channel.value.playlists.error = ''
      const options: Record<string, string> = {}
      if (pageToken) options.pageToken = pageToken
      const response = await fetchChannelPlayListsAPI(channelId, options)
      const currentItems = channel.value.playlists.data[channelId]?.items || []
      channel.value.playlists.data[channelId] = {
        ...response.data,
        items: [...currentItems, ...response.data.items],
      }
      channel.value.playlists.status = AsyncStatus.SUCCESS
    } catch (err: any) {
      channel.value.playlists.status = AsyncStatus.FAIL
      channel.value.playlists.error = err.message
    }
  }

  return {
    channelState: channel,
    fetchChannelProfile,
    fetchChannelVideos,
    fetchChannelPlayLists,
  }
})
