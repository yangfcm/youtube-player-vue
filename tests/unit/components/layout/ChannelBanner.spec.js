import { shallowMount } from "@vue/test-utils";
import { channelIntroResponse } from "../../store/fixtures/channel";
import ChannelBanner from "@/components/layout/ChannelBanner.vue";
import SubscribeButton from "@/components/common/SubscribeButton.vue";

describe("Test ChannelBanner component", () => {
  let wrapper;
  const channelIntro = channelIntroResponse.items[0];
  beforeEach(() => {
    wrapper = shallowMount(ChannelBanner, {
      propsData: {
        channelIntro,
      },
    });
  });
  afterEach(() => {
    wrapper.destroy();
  });

  it("ChannelBanner component matches snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("ChannelBanner component should render correct content", () => {
    expect(wrapper.find("h2").text()).toContain(channelIntro.snippet.title);
    expect(wrapper.find("img").attributes("src")).toBe(
      channelIntro.snippet.thumbnails.medium.url
    );
  });

  it("should contain SubscribeButton component", () => {
    expect(wrapper.findComponent(SubscribeButton).exists()).toBe(true);
  });
});
