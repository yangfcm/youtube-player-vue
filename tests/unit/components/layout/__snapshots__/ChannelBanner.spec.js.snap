// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ChannelBanner component ChannelBanner component matches snapshot 1`] = `
<div class="app-banner-container">
  <div class="app-banner-title-container">
    <div class="app-banner-image"><img src="https://yt3.ggpht.com/a/AATXAJzt3HlB60RSLIwrWgXUfKLZeo0M_rC34JMcM0BPZw=s240-c-k-c0xffffffff-no-rj-mo" alt="BBC Learning English" class="ui tiny image circular"></div>
    <div class="app-banner-title">
      <h2 class="ui header app-banner-title-text">BBC Learning English</h2>
    </div>
  </div>
  <div>
    <app-subscribe-button-stub channelid="UCHaHD477h-FeBbVh9Sh7syA"></app-subscribe-button-stub>
  </div>
</div>
`;
